### DISCUSSION TITLE
Button における `action` の指定方法について検討する。

### DISCUSSION THREAD

**Discussion by YusukeHosonuma** [YusukeHosonuma/Effective-SwiftUI#20] [Ideas]
SwiftUI において、[Button](https://developer.apple.com/documentation/swiftui/button) などの一部のコントロールはタップ時の動作を引数 `action: () -> Void` などで指定できるようになっている。

この時、`action` の書き方は大きく以下の2種類に分けることができる。

1. コードブロックで指定する。
2. 関数への参照を指定する。

```swift
struct ContentView: View {
    @ObservedObject var viewModel = ContentViewModel()
    
    var body: some View {
        // 💡 1. Specify action as code-block.
        Button("hello") {
            viewModel.tapButton()
        }
        
        // 💡 2. Specify action as reference of function.
        Button("hello", action: viewModel.tapButton)
    }
}

class ContentViewModel: ObservableObject {
    func tapButton() {
        // TODO:
    }
}
```

タップ時の処理を特定のメソッド（ここでは`tapButton`）に紐付ける場合、一見すると関数への参照で指定する方が宣言的でスッキリ記述できてベターにも感じる。

しかし、以下のような欠点も存在する。

- （行）ブレークポイントで止めることができない。
- `async`な関数を呼び出したい場合は、結局コードブロックで指定する必要がある。

<img width="545" alt="image" src="https://user-images.githubusercontent.com/2990285/162109513-dfcc3f3f-c3cf-4ef0-944a-193b925c6ae7.png">

デバッガビリティとコードスタイルの統一性の観点から、常にコードブロック形式で指定したほうがベターなケースが多いと予想される。

## Discussion

本項目はトレードオフであり、チーム内で合意が取れるのであればどちらを好んで使用してもよい。

例えば、ブレークポイントについては、（1アクション増えるが）呼び出し先の関数へコードジャンプして、そこにブレークポイントを貼るという方針でも問題ない。

また、Xcode 13 から導入された列ブレークポイント（Column Breakpoint）を利用することもできる（ただし、body 関数の評価時にも止まってしまうので、使い勝手としてはコードブロック版に劣る）。

<img width="497" alt="image" src="https://user-images.githubusercontent.com/2990285/162110335-3ecdbf4d-2a59-4727-8b0a-3e19baf6caf5.png">

`async`な関数へ紐付ける場合のコードスタイルの統一性についても、以下のように自身に転送用の関数を設けることで一貫性を保つことができる。

```swift
struct ContentView: View {
    @ObservedObject var viewModel = ContentViewModel()
    
    var body: some View {
        Button("hello", action: onTapButton)
        Button("async", action: onTapAsyncButton)
    }
    
    func onTapButton() {
        viewModel.tapButton() // 💡
    }
    
    func onTapAsyncButton() {
        Task {
            await viewModel.tapAsyncButton() // 💡
        }
    }
}
```

これらに絶対的な正解はないため、

- デバッガビリティ
- コードスタイルの統一性

の2つの観点から、チーム内で話し合ってルールを決めておくとよい。

**Comment by uhooi** [comment]
私は「ビルド時に警告が出なければ関数の参照で指定（行数がかさばらないため）、警告が出たらコードブロック形式で指定」にしています！
以下のツイートのリプを辿ると有用な情報があるかもしれません :pray:
https://twitter.com/the_uhooi/status/1588550348033380352

**Comment by YusukeHosonuma** [comment]
@uhooi 
> ビルド時に警告が出なければ関数の参照で指定（行数がかさばらないため）、警告が出たらコードブロック形式で指定

以前お会いした時に少し触れたかもしれませんが、私も最近はこの方針にシフトしました 🙏 

しかしながら、実質的に `@MainActor` なのに関数参照で警告が出るのは不便なので、将来的なバージョンで改善を期待したいところだと思うばかりです。

**Reply by uhooi** [reply]
ですね、お話しした内容をもとに書きました笑
改善してほしいですね…。
